File: cpp-sdk/tests/test_esg.cpp
#include <iostream>
#include <cassert>
#include <stdexcept>
#include "erc8040/esg.hpp"

using namespace erc8040;

void test_esg_scoring() {
    ESGScoring scorer;
    ESGScore score = scorer.calculate(85.0, 78.0, 92.0);
    
    assert(score.environmental == 85);
    assert(score.social == 78);
    assert(score.governance == 92);
    assert(score.total >= 80 && score.total <= 90);
    std::cout << "[PASS] test_esg_scoring\n";
}

void test_rating_from_score() {
    assert(ESGScoring::rating_from_score(95) == ESGRating::AAA);
    assert(ESGScoring::rating_from_score(85) == ESGRating::AA);
    assert(ESGScoring::rating_from_score(80) == ESGRating::A);
    assert(ESGScoring::rating_from_score(70) == ESGRating::BBB);
    assert(ESGScoring::rating_from_score(60) == ESGRating::BB);
    assert(ESGScoring::rating_from_score(50) == ESGRating::B);
    assert(ESGScoring::rating_from_score(40) == ESGRating::CCC);
    assert(ESGScoring::rating_from_score(35) == ESGRating::CC);
    assert(ESGScoring::rating_from_score(30) == ESGRating::CC);
    assert(ESGScoring::rating_from_score(20) == ESGRating::C);
    assert(ESGScoring::rating_from_score(5) == ESGRating::D);
    std::cout << "[PASS] test_rating_from_score\n";
}

void test_investment_grade() {
    ESGScoring scorer;
    
    ESGScore high_score = scorer.calculate(90.0, 85.0, 88.0);
    assert(high_score.is_investment_grade() == true);
    
    ESGScore low_score = scorer.calculate(30.0, 25.0, 28.0);
    assert(low_score.is_investment_grade() == false);
    
    std::cout << "[PASS] test_investment_grade\n";
}

void test_rating_to_string() {
    assert(ESGScoring::rating_to_string(ESGRating::AAA) == "AAA");
    assert(ESGScoring::rating_to_string(ESGRating::BB) == "BB");
    assert(ESGScoring::rating_to_string(ESGRating::D) == "D");
    std::cout << "[PASS] test_rating_to_string\n";
}

void test_invalid_weights() {
    try {
        ESGScoring scorer(-1.0, 1.0, 1.0);
        (void)scorer;
        assert(false && "Expected invalid_argument for negative weight");
    } catch (const std::invalid_argument&) {
    }

    try {
        ESGScoring scorer(0.0, 0.0, 0.0);
        (void)scorer;
        assert(false && "Expected invalid_argument for zero weights");
    } catch (const std::invalid_argument&) {
    }

    std::cout << "[PASS] test_invalid_weights\n";
}

int main() {
    std::cout << "=== ERC-8040 C++ SDK - ESG Tests ===\n\n";
    
    test_esg_scoring();
    test_rating_from_score();
    test_investment_grade();
    test_rating_to_string();
    test_invalid_weights();
    
    std::cout << "\n✅ All ESG tests passed!\n";
    return 0;
}

File: python-sdk/tests/test_esg.py
"""Tests for ESG scoring and rating functionality."""

from erc8040_sdk import ESGRating, ESGScore, ESGScoring


def test_esg_rating_from_score():
    """Test ESG rating calculation from score."""
    assert ESGRating.from_score(95) == ESGRating.AAA
    assert ESGRating.from_score(75) == ESGRating.BBB
    assert ESGRating.from_score(45) == ESGRating.CCC


def test_esg_rating_investment_grade():
    """Test investment grade classification."""
    assert ESGRating.AAA.is_investment_grade
    assert ESGRating.BBB.is_investment_grade
    assert not ESGRating.BB.is_investment_grade


def test_esg_scoring():
    """Test ESG scoring calculation."""
    scoring = ESGScoring()
    score = scoring.calculate(80.0, 70.0, 90.0)
    assert score.total > 0
    assert score.rating in list(ESGRating)


def test_esg_score_create():
    """Test ESG score creation."""
    score = ESGScore.create(90.0, 85.0, 80.0)
    assert score.environmental == 90.0
    assert score.social == 85.0
    assert score.governance == 80.0
    assert abs(score.total - 85.0) < 0.01
    assert score.rating == ESGRating.AA


def test_esg_scoring_custom_weights():
    """Test ESG scoring with custom weights."""
    scoring = ESGScoring(environmental_weight=2.0, social_weight=1.0, governance_weight=1.0)
    score = scoring.calculate(80.0, 60.0, 60.0)
    # (80*0.5 + 60*0.25 + 60*0.25) = 70
    assert abs(score.total - 70.0) < 0.01


def test_esg_score_is_investment_grade():
    """Test investment grade check on ESG score."""
    score = ESGScore.create(90.0, 85.0, 80.0)
    assert score.is_investment_grade()

    low_score = ESGScore.create(60.0, 55.0, 50.0)
    assert not low_score.is_investment_grade()
